lib/_tls_wrap.js

```diff
@@ -426,7 +426,11 @@ TLSSocket.prototype._init = function(socket, wrap) {
 
     // Destroy socket if error happened before handshake's finish
     if (!self._secureEstablished) {
-      self.destroy(self._tlsError(err));
+      if (!self._controlReleased) {
+        self.destroy(err);
+      } else {
+        self.destroy(self._tlsError(err));
+      }
```

Comment:
@indutny `self._tlsError` returning `null` is exactly what i am addressing here. When handshake fails and `ssl.onerror` is called control is not released yet, so `self.destroy` is called with `null`, and `error` event is not fired in `destroy`.

Here is debug output from upstream:
```
NET 26944: listen2 null 1337 4 false undefined
NET 26944: _listen2: create a handle
NET 26944: bind to ::
NET 26944: onconnection
NET 26944: _read
NET 26944: Socket._read readStart
NET 26944: _read
TLS 26944: onhandshakestart
NET 26944: destroy null
NET 26944: destroy
NET 26944: close
NET 26944: close handle
NET 26944: destroy undefined
NET 26944: destroy
NET 26944: close
NET 26944: close handle
NET 26944: has server
NET 26944: SERVER _emitCloseIfDrained
NET 26944: SERVER handle? true   connections? 0
NET 26944: emit close
NET 26944: emit close
```
...and here after the patch:

```
NET 31249: listen2 null 1337 4 0 undefined
NET 31249: _listen2: create a handle
NET 31249: bind to ::
NET 31249: onconnection
NET 31249: _read
NET 31249: Socket._read readStart
NET 31249: _read
TLS 31249: onhandshakestart
NET 31249: destroy Error: 140507750061888:error:1408A0C1:SSL routines:ssl3_get_client_hello:no shared cipher:../deps/openssl/openssl/ssl/s3_srvr.c:1418:

NET 31249: destroy
NET 31249: close
NET 31249: close handle
NET 31249: destroy undefined
NET 31249: destroy
NET 31249: close
NET 31249: close handle
NET 31249: has server
NET 31249: SERVER _emitCloseIfDrained
NET 31249: SERVER handle? true   connections? 0
NET 31249: emit close
NET 31249: emit close
```
Now, just replacing `self.destroy(self._tlsError(err))` with `self.destroy(err)` there also fixes #8803 and same tests pass, but since `self._tlsError` emits additional `'_tlsError'` event which could be used in something not covered with tests, i opted for safer approach.